### ISSUE TITLE
Performance comparison with fastText

### ISSUE THREAD

**Issue by kperi** [facebookresearch/StarSpace#132]
Hi, 
a note on performance,

I am testing the same input file with fastText & StarSpace,

fastText  running time is ~45 mins with params:

```
./fasttext cbow -minCount 100 -dim 100  -input input.txt.txt -output cbow_mincount_dim_100
````

StarSpace running time is ~5hrs with params:

```
./starspace train -trainFile input.txt -model sp_100_10 -trainMode 5 -dim 10 -minCount 100 -loss hinge -verbose true
```

Input file in lines is 15.2m, vocabulary size with minCount 100 is 95K words

Is this expected behaviour or am I missing something obvious here?

Thanks!

**Comment by mattyd2** [comment]
Hi @kperi,

I've noticed the same thing with some recent experiments I've been running. 

The FastText training time is much less for the same input file with the same parameters.

**Comment by jaseweston** [comment]
Hi! You can't really compare them for the same parameters because they have different algorithms underneath which use different parameters. You should really compare them for the same amount of training time, by test error (whichever parameters are best for each), you can control training time in StarSpace with the limit on it using the flag 'maxTrainTime'. Or else look for the best test error for each method, and then how long it took (again, independent of parameters).
Having said that for classification I would say fastText _is_ faster. There are some comparisons in our paper: https://arxiv.org/abs/1709.03856.
The real bonus of StarSpace is using it for other tasks, e.g. where the thing you what to output has features, also in the paper.
We are working on looking into new ideas to make StarSpace for all cases, however.

**Comment by kperi** [comment]
Ηι @jaseweston,

Thanks for the reply,
I am using word2vec, fastText, GloVe and StarSpace  for embeddings generation, to be used as features in ML models. 
 
The difference in the average training time between the 3 first and StarSpace is significant on the same training set (and also the fact that the default dim for StarSpace is 10, whereas in fastText/glove 100 is usually a good place to start).

I understand that they are different algorithms (and I haven't have closer loo on implementation details yet), so my question was mostly if this is expected behaviour

Thanks very much,
Kostas

**Comment by jaseweston** [comment]
@ledw  we should change the default dim to be bigger than 10 as that isn't good for a lot of problems, and adjust any examples expecting 10 so they still behave the same
I would say speed-wise, it's probably not really as much to do with the embedding dimension (although that does change it of course) as with the parameters 'maxNegSamples', 'negSearchLimit', 'epoch' and 'lr'

**Comment by ledw** [comment]
@kperi Thanks for the suggestion. In addition to what @jaseweston mentioned, it is also known that hinge loss is slower than the softmax loss. You can try trainMode 5 with the softmax loss. Hinge loss does give better results on other ranking tasks so we set it as default.
For the default dimension, I'll change it to 100.

**Comment by ledw** [comment]
@kperi have you tried dim 100 and/or softmax loss on your data? Any updates? Thanks.

**Comment by kperi** [comment]
Hi @ledw 

I am getting somewhere: 

```
lr: 0.01
dim: 100
epoch: 5
maxTrainTime: 8640000
saveEveryEpoch: 0
*loss: softmax*
margin: 0.05
similarity: cosine
maxNegSamples: 10
*negSearchLimit: 10*
thread: 10
minCount: 100
minCountLabel: 1
label: __label__
ngrams: 1
bucket: 2000000
*adagrad: 0*
trainMode: 5
fileFormat: fastText
normalizeText: 0
dropoutLHS: 0
dropoutRHS: 0
Start to initialize starspace model.
Build dict from input file : input.txt
Read 195M words
Number of words in dictionary:  94997
Number of labels in dictionary: 0
Loading data from file : input.txt
Total number of examples loaded : 14849477
Initialized model weights. Model size :
matrix : 94997 10
Training epoch 0: 0.01 0.002
*Epoch: 0.4%  lr: 0.009992  loss: 2.116200  eta: 12h10m  tot: 0h0m40s  (0.1%)

```

versus:


```
lr: 0.01
*dim: 10*
epoch: 5
maxTrainTime: 8640000
saveEveryEpoch: 0
*loss: softmax*
margin: 0.05
similarity: cosine
maxNegSamples: 10
*negSearchLimit: 10*
thread: 10
minCount: 100
minCountLabel: 1
label: __label__
ngrams: 1
bucket: 2000000
*adagrad: 0*
trainMode: 5
fileFormat: fastText
normalizeText: 0
dropoutLHS: 0
dropoutRHS: 0
Start to initialize starspace model.
Build dict from input file : input.txt
Read 195M words
Number of words in dictionary:  94997
Number of labels in dictionary: 0
Loading data from file : input.txt
Total number of examples loaded : 14849477
Initialized model weights. Model size :
matrix : 94997 10
Training epoch 0: 0.01 0.002
*Epoch: 6.0%  lr: 0.009877  loss: 2.060369  eta: 2h11m  tot: 0h1m36s  (1.2%)
````


so from ```dim=10``` to ```dim=100`` using softmax multiplies estimated time by a factor of 6.

**Comment by ledw** [comment]
@kperi Hi, thanks for reporting. It looks a bit strange to me that in your 100 dim config, the matrix size is still 94997 * 10 (it should be 94997 * 100). In addition, the estimate of running time at the beginning may not be accurate and it will change a bit after training for a while. Also the negSearchLimit is something you can tune for speed: you can try to reduce it to 5 or something.

**Comment by kperi** [comment]
@ledw most probably I messed up somewhere with copying and pasting params. is negSearchLimit related with beam size/beam search of some kind? 

Anyway, I will surely try a few more times with different configs. 


Thanks!

**Comment by Tiriar** [comment]
I still find it quite weird that the training times of StarSpace and FastText differ so much. After reading the respective papers, I believe that with using StarSpace with softmax and dot product as a similarity function, their objective functions (considering the SkipGram negative sampling version of FastText) should be basically the same.

When I train both on the same data (corpus with 456 mil. words / 2.2 mil. words as vocabulary) while matching the settings as best as I could (minCount 1, dim 300, 2 epochs, 10 negatives, 10 threads), FastText takes cca 560 minutes of CPU time to train, while StarSpace takes over 15000 minutes of CPU time.

One error I made in the comparison is that I did not match the learning rates. Other thing is that I believe that FastText uses regular SGD, while StarSpace uses AdaGrad (which is apparently slower for smaller number of epochs), but I still believe that I must be missing something in the way StarSpace learns that makes it so much slower...

(I am working on a thesis about word embeddings and the thing is that I feel compelled to give a good reason for the difference in training times. I would usually just write something about the implementations being different, but with such a big difference, I guess I must be missing something about the algorithm itself :D)

**Comment by ledw** [comment]
@Tiriar Thanks for reporting. As @jaseweston mentioned, we have not been optimized starspace for training word level embeddings. I looked into the word level embedding training part (train mode 5) and compared with fastText, with the following find-outs:
1. There are several implementation detail differences: fastText discards word in each example according to its frequency, while starspace takes each word in each example when doing the sliding window;  fastText also set a random sliding window size uniformly from [1, ws] (where ws is in the argument), while starspace just use ws as sliding window size. 
2. When using a similar dataset as you described and run starspace and fastText with similar parameters (minCount, dim, epochs, negatives (the parameter for #negatives is `-negSearchLimit` in starspace)) and 40 threads, I got starspace ~7x slower than fastText. However, I found the speed difference is much larger with 10 threads (namely fastText runs about the same with 10 vs 40 threads, while starspace runtime is roughly linear to number of threads). I'm not sure what caused this, and will investigate more.

In general we will do more optimization on starspace algorithms with respect to the negative sampling part and will keep you updated. Again thanks for experimenting!

**Comment by Tiriar** [comment]
@jaseweston @ledw Thank you for your replies, especially the point with word discarding and context window is interesting. I can also try running the algorithms on more threads as you suggested.

I gotta say, even though you did not focus on word embeddings, the models I trained with StarSpace outperform all other embedding algorithms that I tried to train on the dataset I described in my previous post (of course I am comparing those by training runtime) - I tested the performance on the STS Benchmark dataset. Also while the performance of word2vec and sent2vec tends to drop significantly when reducing the size of the training data, StarSpace is able to keep its performance almost as well as FastText...

You created one hell of an embedding algorithm :+1: :)

**Comment by jaseweston** [comment]
@Tiriar  Cool! we'd love to see any results you could share!

**Comment by ledw** [comment]
@Tiriar Thanks! That's great to know. Yes we'd love to see anything you can share from your experiments and analysis. That could help us to make the algorithm better. :-)

**Comment by Tiriar** [comment]
@jaseweston @ledw Sure, it will all be in my thesis, which will be publicly available after I finish it (in cca a month), so I will share the link with you ;-).

The goal of my thesis is to create an intent recognition system that uses word / sentence embeddings - and most of the work went into comparing different embedding algorithms in terms of performance, HW requirements and training properties on different datasets. It also explores some embedding compression methods.

I will be glad if it could help someone.

**Comment by Tiriar** [comment]
@jaseweston @ledw Hello, if you are still interested, the thesis is available here:
http://hdl.handle.net/10467/77029

The comparison of the embedding algorithms is mostly reported in Chapter 3. We want to make a paper from the results, since no one really reads university theses, but this one is quite short :-).

**Comment by jaseweston** [comment]
nice! thanks!
